### PR TITLE
Use json for i18n

### PR DESCRIPTION
**Pull request by Hamms**
Updates Dashboard and all our i18n processing scripts to use json rather than yml for storage of the i18n data so we can stop dealing with all the issues arising from yaml complexity and inconsistency.

The actual conversion of the data is handled in https://github.com/code-dot-org/code-dot-org/pull/26310

Depends on https://github.com/svenfuchs/i18n/pull/429

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -13,11 +13,18 @@ end
...
 # Ref: https://github.com/svenfuchs/i18n/pull/423
 # Support numeric keys in Simple backend.
 # Ref: https://github.com/svenfuchs/i18n/pull/422
-gem 'i18n', github: 'wjordan/i18n', branch: 'cdo'
+# Add JSON backend support
+# Ref: https://github.com/svenfuchs/i18n/pull/429
+#
+# This can be updated to a regular versioned dependency as soon as a version
+# including these features is released
+gem 'i18n', github: 'svenfuchs/i18n', ref: '40bca10'
```
Comment (Hamms) [inline Gemfile]: I'm not sure how I feel about the relative disadvantages of depending on a specific commit in the official repo vs a branch in our custom fork, but until a new version of the gem is released we have to go with one of those.

Another alternative would be to create a `code-dot-org/i18n-gem` repo in lieu of Will's personal fork.

I'm open to any and all thoughts on the best way to go here

**Comment by wjordan** [comment]
> so we can stop dealing with all the issues arising from yaml complexity and inconsistency.

Maybe the discussion details are somewhere else but can you elaborate on the issues? Are there any currently-unresolved issues this PR will fix, or is this defending against further issues we expect are likely arise based on ones we've already dealt with?

My concern is that there's a lot of data-migration effort and risk of breaking things involved in making this conversion, so I want to be more clear what the benefits are.

**Comment by Hamms** [comment]
Sure!

Most of the problems with YAML come from its complexity, and the inconsistency with which its various features are implemented by various systems. We've already had to do [quite a bit of work](https://github.com/code-dot-org/code-dot-org/blob/2a8f601f66bb885627a0bcc6ddbd1f990ed5621c/bin/i18n/i18n_script_utils.rb#L13-L40) to deal with inconsistencies between the Ruby version of YAML and whatever version Crowdin uses, and there are a couple more inconsistencies that we're already aware of but have not yet fixed (support for complex mapping syntax and for consistent indentation syntax), and I fully expect there to be yet more issues of that form that we have yet to encounter.

**Comment by Hamms** [comment]
@wjordan @joshlory @nkiruka PTAL

**Review by nkiruka — APPROVED** [review]
Great!

**Comment by joshlory** [comment]
Removing myself from the reviewers list — I think @wjordan is the main stakeholder here for sign off!

**Review by wjordan — COMMENTED** [review]
Not knowing further details on the 'complex mapping syntax' and 'consistent indentation syntax' issues, I can't say I'm convinced the complexity/inconsistency argument alone makes this conversion necessary. Considering Crowdin lists 'Ruby on Rails & YAML' as a [supported format](https://support.crowdin.com/supported-formats/), I'd expect Crowdin to own and fix any related bugs in its support for this format.

I can imagine other additional reasons to migrate i18n data to JSON, loading/parsing performance being one obvious reason (though it might not be the _best_ performing data format, it should be better than YAML). Future plans to easily share translation files across different platforms (e.g., sharing i18n across client-side JavaScript code) is another. There's a human-readability tradeoff to consider as well (for which I would favor YAML, though others more familiar with JSON might favor that instead), but if we make the assumption that these i18n files no longer being human-read and are instead being consumed by Crowdin exclusively, then this is less of a factor.

Anyway, since there are clear reasons behind this migration my remaining hesitation is below the level of blocking concern, so feel free to go ahead with this :+1:

**Comment by Hamms** [comment]
> Considering Crowdin lists 'Ruby on Rails & YAML' as a supported format, I'd expect Crowdin to own and fix any related bugs in its support for this format.

Oh, I might have been unclear; the problem isn't with _bugs_ in the various yaml parsers, it's with inconsistencies in those parsers that arise from different (but equally valid) interpretations of the yaml spec.

As as aside, of the various problems I've encountered the one that I'm most inclined to call a bug is actually [this line](https://github.com/yaml/libyaml/blob/85d1f168ef39f4c4419ef362d849ea225c53fffc/src/emitter.c#L1161) in libyaml that forces any string with a length greater than 128 characters to use the complex mapping syntax rather than the simple mapping syntax, a restriction I cannot find any basis for in the yaml spec itself. (and one which Crowdin doesn't support, but which Psych - our parser - absolutely requires).

I agree that YAML is definitely a better human-readable format, but these files are definitely optimized for machine consumption, particularly because most of them are actually computer-generated in the first place rather than being human-written at any point in their lifecycle.

Thanks for the careful consideration!

File: Gemfile [hunk 2]
```diff
@@ -13,11 +13,7 @@ end
 gem 'rails', '~> 5.0.1'
 gem 'rails-controller-testing'
 
-# Add CacheFile backend module.
-# Ref: https://github.com/svenfuchs/i18n/pull/423
-# Support numeric keys in Simple backend.
-# Ref: https://github.com/svenfuchs/i18n/pull/422
-gem 'i18n', github: 'wjordan/i18n', branch: 'cdo'
+gem 'i18n', '1.3.0'
```
Comment (wjordan) [inline Gemfile:16]: (forgot to mention) thanks for updating this back to the latest published gem! :+1: :tada:

**Pull request closed without merging**